### PR TITLE
Fix several cases of relic_data failing to update properly

### PR DESCRIPTION
**Pull request by chaosvolt**
#### Summary

SUMMARY: Bugfixes "Fix several functions that did not update enchantment cache and sight limits where needed"

#### Purpose of change

This is a bundle of several fixes to the myriad instances of enchantment/relic effects not updating correctly, causing it to require a turn to kick in properly.

Also ports over part of a fix in DDA for relic_data failing to work without a delay when game is loaded. Most of the rest of their PR involves changes to code that's diverged between DDA and BN, so for the rest I just went on a bug-hunting spree myself, generally checking functions that call `reset_encumbrance` is a good idea.

#### Describe the solution

1. In activity_handlers.cpp, set `activity_handlers::spellcasting_finish` to call `recalculate_enchantment_cache` since a spell might do tihngs like create enchanted items.
2. In bionics.cpp, I noticed that `Character::activate_bionic` lacked a call to `recalculate_enchantment_cache` like all the other relevant functions in it. This fixes issues where turning a bionic on fails to provide active enchantment effects until after a turn later, for example Cloaking System failing to kick in.
3. In character.cpp, added calls to `recalculate_enchantment_cache` to `Character::wear_item` and `Character::takeoff`, to fix cases of relic data failing to update when wearing or removing enchanted gear.
4. In character.h, moved `enchantment_cache` to public section of the relevant area, in order for other functions to use it when needed.
5. In game.cpp, ported over Ramza's addition of enchantment update functions placed within `game::load`, which fixes relic data stuff taking a turn to kick in after you load up a game. 
6. In iuse_actor.cpp, added calls toboth `recalculate_enchantment_cache` and `recalc_sight_limits` to `iuse_transform::use` and `salvage_actor::cut_up` as those may change or destroy relics. The former is the more important of the two, since transforming relics are a recurring source of fuckiness. This also ensures that changes to vision range from item transformation, or addition of clairvoyance from relics, will kick in properly. And in addition, an application of the activate_passive function was needed for transforming items to correctly get all enchantment effects to kick in on transformation, just the recalcs weren't enough.

#### Testing

1. Set up a save holding an active cube of shame.
2. Compiled and load-tested.
3. Confirmed the stat effects of the cube were visible immediately, instead of requiring me to wait a turn later.
4. Installed Cloaking System, confirmed the effect now shows up on the `@` screen immediately after being turned on, and not a turn afterward.
5. Temporarily added Arcana to mods, tested that the clairvoyance from a mask of insight takes effect immediately.
6. Also tested that the stat boosts from the cursed blade activate the instant you turn it on.

Problem: it still doesn't correctly update stuff when you turn a transformed item OFF, and I'm not sure why?

#### Additional context

Related PR, by @Ramza13: https://github.com/CleverRaven/Cataclysm-DDA/pull/61418

### CODE REVIEW

File: src/character.h [hunk 1]
```diff
@@ -2300,6 +2297,10 @@ class Character : public Creature, public visitable<Character>
         std::optional<double> get_npc_ai_info_cache( npc_ai_info key ) const;
 
         safe_reference<Character> get_safe_reference();
+
+        // a cache of all active enchantment values.
+        // is recalculated every turn in Character::recalculate_enchantment_cache
+        pimpl<enchantment> enchantment_cache;
```
Comment (Coolthulhu) [inline src/character.h:2303]: Evil code warning.
If you need to make something public just to call "recalculate this thing", it's probably not being done well.

From what I can tell, you only use it once, to call `activate_passive`. Can you do it some other way? Why is it not done in cache recalculation itself?

Comment (chaosvolt) [reply src/character.h:2303]: > it's probably not being done well.

That's spell and relic code in a nutshell right there. I'm not sure why it needs that to function nor how to make it suck less. :/

Comment (Coolthulhu) [reply src/character.h:2303]: What happens if you only recalculate the cache and don't activate passives manually?
It sounds like the passives are proced at start of the turn and some of the effects happen after that. Just procing passives twice could double some modifiers, if they were present before and after the second proc.

File: src/game.cpp [hunk 2]
```diff
@@ -2621,6 +2622,8 @@ bool game::load( const save_t &name )
     calendar::set_season_length( ::get_option<int>( "SEASON_LENGTH" ) );
 
     u.reset();
+    u.recalculate_enchantment_cache();
+    u.enchantment_cache->activate_passive( u );
```
Comment (olanti-p) [inline src/game.cpp:2626]: I don't know how it works in DDA, but in BN `activate_passive()` not only applies stat and dodge bonuses cleared by `u.reset()` 2 lines above, but also fires off various intermittent activation effects and adds 1 second to all effects caused by enchantments, so loading the game can randomly cause the relic to fire off its intermittent effects, or stack effects for longer durations (you can reproduce it by wielding architect's cube and repeatedly saving and loading the game).

I think `activate_passive()` can be split in 2 parts: one that applies bonuses and should be called here, and one that applies effects and intermittent activation and should _not_ be called here.

File: src/character.h (hunk 1 above)
Comment (chaosvolt) [reply src/character.h:2307]: Ah, hmm. Not sure what'd be the best solution for that.

File: src/game.cpp (hunk 2 above)
Comment (chaosvolt) [reply src/game.cpp:2626]: Oh hecc, good idea probably. Might try and work on that later today I guess, though given this is proving clunky to fix I might be better off closing it and messing with it another time, especially given still juggling some stuff IRL lately.

**Comment by chaosvolt** [comment]
Gonna go ahead and close for now as I'm not sure how to really hack my way through a better fix for this.

**Pull request closed without merging**